### ISSUE TITLE
The "LocalOnly" parameter is not supported by the "GetApplicationParametersFiles" task. When using VSTS hosted build agents.

### ISSUE THREAD

**Issue by MikkelHegn** [Azure/service-fabric-issues#51]
When trying to build an SF project that has been upgraded to 5.2 with the hosted agents we get the following error:
"C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\Service Fabric Tools\Microsoft.VisualStudio.Azure.Fabric.ApplicationProject.targets (116, 36)
The "LocalOnly" parameter is not supported by the "GetApplicationParametersFiles" task. Verify the parameter exists on the task, and it is a settable public instance property."

"C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\Service Fabric Tools\Microsoft.VisualStudio.Azure.Fabric.ApplicationProject.targets (115, 5)
The "GetApplicationParametersFiles" task could not be initialized with its input parameters."

**Comment by MikkelHegn** [comment]
We have a backward compatibility issue with the build agents in VSTS, fix is on it;s way - check back for an ETA.

**Comment by MikkelHegn** [comment]
We have published an updated version of the Microsoft.VisualStudio.Azure.Fabric.MSBuild package - 1.2.1.
This version fixes the backward compatibility issue, however as Service Fabric Application projects are already updated by the tooling, there is additional work to do as a workaround for now.

After having updated the NuGet package, you need to manually change the references in the .sfproj project file of the Service Fabric Application project.

1. **Right-click** the Service Fabric Application project in Visual Studio Solution Explorer and choose **Unload Project**.
2. **Right-click** the Service Fabric Application project** in Visual Studio Solution Explorer and choose **Edit xxx.sfproj**
3. Change the four (4) references to the Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.0 to Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1. All references are in two (2) lines, both starting with an **Import** tag. **Do not change anything else in the file!**
4. Save and close the file
5. **Right-click** the Service Fabric Application project** in Visual Studio Solution Explorer and choose **Reload Project**

Your project should now be able to build on VSTS hosted build agents again.

As soon as the hosted build agents get updated, the above work around is no longer needed. Please post any issues you encounter here.
We will update this issue as soon as we have a permanent fix.

**Comment by BlueBasher** [comment]
Where can I find this 1.2.1 packages? I can't find it in the official NuGet feed.
Or should I use a separate NuGet feed for this?

Thanks,
Marco

Date: Mon, 19 Sep 2016 21:50:59 -0700

**Comment by Javietto** [comment]
Hello,

After applying said fix to use MSBuild 1.2.1 I get a new error message:

C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets (2482, 7)
C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(2482,7): Error MSB4057: The target "CreateManifestResourceNames" does not exist in the project.

That is when running on VSTS hosted agent, on a private agent it works just fine (even prior to the workaround).

On an unrelated note, is there a github for VSTS team? Do we know when they will update nuget.exe version for hosted agents?

Thank you.

**Comment by andersosthus** [comment]
The fix with MSBuild 1.2.1 worked great for us. Builds now working on the hosted agents again.

**Comment by Javietto** [comment]
Thanks to andersosthus comment I've double checked my worked and noticed I've only changed half of the references. Sorry about my previous comment.

**Comment by mburumaxwell** [comment]
I have found that if I change the reference before I install the NuGet package into the ServiceFabric application project, VS keep saying that it wants to restore packages but it does not work. After installation and updating the references, everything else is okay.

**Comment by Javietto** [comment]
@mburumaxwell Make sure you also change the reference in packages.config of our sfproject to use 1.2.1

**Comment by MikkelHegn** [comment]
Hi All,

There is a Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1 package out there, unlisted, but you are able to pick it up.

This version fixes the backward compatibility issue, however as Service Fabric Application projects are already updated by the tooling, there is additional work to do as a workaround for now.

This workaround will unblock you, if you are building using VSTS hosted build agents, for all projects upgraded to version 1.2, as well as new project created using 1.2 of the tooling. The permanent fix will be an update to the hosted build agents, which we will post an update on as soon as we have an ETA.

To start using the 1.2.1 package:

1. Change the version of the package in the package.json file of your Service Fabric Application Project. Package reference should be: Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1.

2. **Right-click the Service Fabric Application project** in Visual Studio Solution Explorer and choose **Unload Project**.

3. **Right-click the Service Fabric Application project** in Visual Studio Solution Explorer and choose **Edit xxx.sfproj**.

4. Change the four (4) references to the Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.0 to Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1. All references are in two (2) lines, both starting with an Import tag. Do not change anything else in the file!

5. **Save** and **close** the file

6. **Right-click the Service Fabric Application project** in Visual Studio Solution Explorer and choose Reload Project

Your project should now be able to build on VSTS hosted build agents again.

Please post any issues you encounter here.

**Comment by andersosthus** [comment]
Just wanted to say, for us we had to update the ```packages.config``` file in our Service Fabric Application project, not ```package.json```

**Comment by MikkelHegn** [comment]
Thanks Anders - just updated the instructions above...

**Comment by ChristianJensen** [comment]
After applying the workaround above, I'm getting another error. 
C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(724,5): Error : The OutputPath property is not set for project 'foo.sfproj'. Please check to make sure that you have specified a valid combination of Configuration and Platform for this project. Configuration='Release' Platform='x64'. You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project.

Builds locally, but not in VSTS using hosted build agent. All NuGet packages are excluded via .gitignore.

**Comment by MikkelHegn** [comment]
Hi Christian,

I've see this problem if you are using .NET core and any of your projects do not have an Release|x64 build configuration.

**Comment by ChristianJensen** [comment]
Hi Mikkel, 

We are not using .NET Core in our solution and we do have a Release|x64 build configurations for all projects. 

I found a blog post with a fix for the same error message, but different cause. The issue in this case seems to be an incorrect version of the Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1\Microsoft.VisualStudio.Azure.Fabric.Application.targets file. We have excluded NuGet packages from source control as well as using a NuGet package restore build task, but could it be that the build sub folder is not correctly restored on the VSTS hosted build agent and thus missign the correct targets file. Anyway to diagnose the content of that folder after build?

Snippet from our .gitignore:

# NuGet Packages
*.nupkg
# The packages folder can be ignored because of Package Restore
**/packages/*
# except build/, which is used as an MSBuild target.
!**/packages/build/
# Uncomment if necessary however generally it will be regenerated when needed
#!**/packages/repositories.config
# NuGet v3's project.json files produces more ignoreable files
*.nuget.props
*.nuget.targets

# Microsoft Azure Build Output
csx/
*.build.csdef

**Comment by Javietto** [comment]
@ChristianJensen I was having the same issue as you guys and turns out I did not properly update the sfproj. There are **2** <Import> lines that contain references to the msbuild version used by the project, one at the top and one at the bottom, you need to update them both.

You might have done this step correctly already, just double check, mistakes happen.

**Comment by ChristianJensen** [comment]
@Javietto I have reviewed my changes to the .sfproj file and it looks good to me, but I'm having a second set of eyes on the changes now. :)

**Comment by lapellaniz** [comment]
I am running into the same issue as @ChristianJensen. I updated the project file to specify 1.2.1 vs 1.2.0 and now I get the following:

"2016-09-21T16:34:22.9753064Z ##[error]C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(724,5): Error : The OutputPath property is not set for project 'Platform.FeatureFlags.sfproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='release'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project."

I verified that we are building using x64 and Release. We are using a .NET Core project in our case.

**Comment by MikkelHegn** [comment]
In my case, it was a dotnet core class library build configuration, that was set to AnyCPU, and not specifically a configuration named x64. I think I had to somehow trick this change through editing the solution file directly.

This is what worked for me.

From the class library .csproj (I remember from earlier conversations - that this is the silver bullet here):
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
    <PlatformTarget>x64</PlatformTarget>
    <OutputPath>bin\x64\Debug\</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
    <PlatformTarget>x64</PlatformTarget>
    <OutputPath>bin\x64\Release\</OutputPath>
  </PropertyGroup>

**Comment by ChristianJensen** [comment]
This is our .sfproj file. Project refs removed...

`<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="..\packages\Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1\build\Microsoft.VisualStudio.Azure.Fabric.Application.props" Condition="Exists('..\packages\Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1\build\Microsoft.VisualStudio.Azure.Fabric.Application.props')" />
  <PropertyGroup Label="Globals">
    <ProjectGuid>541fc1e5-a070-4f49-824a-a7b5238182f1</ProjectGuid>
    <ProjectVersion>1.2</ProjectVersion>
  </PropertyGroup>
  <ItemGroup Label="ProjectConfigurations">
    <ProjectConfiguration Include="Debug|x64">
      <Configuration>Debug</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Release|x64">
      <Configuration>Release</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
  </ItemGroup>
  <ItemGroup>
    <None Include="ApplicationPackageRoot\ApplicationManifest.xml" />
    <None Include="ApplicationParameters\Cloud.xml" />
    <None Include="ApplicationParameters\Local.1Node.xml" />
    <None Include="ApplicationParameters\Local.5Node.xml" />
    <None Include="PublishProfiles\Cloud.xml" />
    <None Include="PublishProfiles\Local.1Node.xml" />
    <None Include="PublishProfiles\Local.5Node.xml" />
    <None Include="Scripts\Deploy-FabricApplication.ps1" />
  </ItemGroup>
  <ItemGroup>
    <Content Include="packages.config" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="*removed*" />
    <ProjectReference Include="*removed*" />
    <ProjectReference Include="*removed*" />
    <ProjectReference Include="*removed*" />
    <ProjectReference Include="*removed*" />
    <ProjectReference Include="*removed*" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.Common.targets" />
  <PropertyGroup>
    <ApplicationProjectTargetsPath>$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\Service Fabric Tools\Microsoft.VisualStudio.Azure.Fabric.ApplicationProject.targets</ApplicationProjectTargetsPath>
  </PropertyGroup>
  <Import Project="$(ApplicationProjectTargetsPath)" Condition="Exists('$(ApplicationProjectTargetsPath)')" />
  <Import Project="..\packages\Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1\build\Microsoft.VisualStudio.Azure.Fabric.Application.targets" Condition="Exists('..\packages\Microsoft.VisualStudio.Azure.Fabric.MSBuild.1.2.1\build\Microsoft.VisualStudio.Azure.Fabric.Application.targets')" />
</Project>`

This is our packages.config file: 

`<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Microsoft.VisualStudio.Azure.Fabric.MSBuild" version="1.2.1" />
</packages>`

When following the workaround process above NuGet packages were restored by right-click on solution and selecting "Restore NuGet packages" 

No .NET Core projects in use. All projects have build configuration  of Release|x64. 

Build still not working using hosted build agent in VSTS. 

What are the options here? What's normally the turnaround for a refresh of hosted build agents to support the new AF SDK out of the box?

**Comment by lapellaniz** [comment]
I was able to get passed this!! 

I confirmed that Release and x64 were set in the VSTS build tasks for the solution and the service fabric project itself. I updated the .sfproj and specified 1.2.1 for the version of MSBuild. If you do a replace there are 4 entries that have it in this file. I also updated the packages,config for that project and specified 1,2,1 <package id="Microsoft.VisualStudio.Azure.Fabric.MSBuild" version="1.2.1" />.

Our solution has .sfproj and xproj so there is no place to specify the outputpath for those.

**Comment by ChristianJensen** [comment]
Still facing this issue, driving me insane. :( No CI build and no CD release = the patient is dead.

**Comment by MikkelHegn** [comment]
Hi Christian and lapalleniz,

At least we have one working...

Christian, can you share the logs from the hosted agent? mikhegn at microsoft dot com?

**Comment by ChristianJensen** [comment]
Hi Mikkel,

The issue has been fixed now, turned out to be a wildcard on the sfproj build step which picked up the wrong .sfproj file. An old sfproj file had been added to the source at the same time of the SDK upgrade by a team member. Obviously this sfproj file was not using 1.2.1. Thanks again for your support!

**Comment by MikkelHegn** [comment]
Hi Christian, Good to hear you got it fixed.
